:x-pack:qa:rolling-upgrade:v6.8.10#upgradedClusterTest: IO error while waiting cluster

**Build scan**: https://gradle-enterprise.elastic.co/s/yvd4eyecgtota

**Repro line**: `./gradlew -p x-pack/qa/rolling-upgrade check`

**Reproduces locally?**: yes. wait. no. I'm confused.

**Applicable branches**: 7.x

**Failure history**:
I don't see it in build-stats but I'm not good at searching. I've seen it several times today.

**Failure excerpt**:
```
»  Caused by: org.elasticsearch.transport.ActionNotFoundTransportException: No handler for action [indices:admin/data_stream/delete]
```
...

```
»  Caused by: org.elasticsearch.ElasticsearchException: node doesn't have meta data for index [my_old_index/Vn7jbwk0TF6Bh9MubpGT0Q]
»  	at org.elasticsearch.indices.store.TransportNodesListShardStoreMetadata.listStoreMetadata(TransportNodesListShardStoreMetadata.java:160) ~[?:?]
```

The `ActionNotFoundTransportException` is expected to be printed in the logs. After each test data streams get cleaned. If a node doesn't support data streams then that request fails with either a 405 or a 500 (action not found error). This is has been taken into account when [executing the delete data stream api.](https://github.com/elastic/elasticsearch/blob/7.x/test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java#L661)

I think the reason that this test fails is that at some stage during the rolling upgrade the cluster is unable to elect a new master node returning health response:

```
[2020-06-02T19:24:45,332][WARN ][o.e.c.c.ClusterFormationFailureHelper] [v6.8.10-2] master not discovered yet, this node has not previously joined a bootstrapped (v7+) cluster, and [cluster.initial_master_nodes] is empty on this node: have discovered [{v6.8.10-2}{VOcLb7ZATs2T0VZfffUBEw}{FIQSWqWYRHG2-5gdHMidwA}{127.0.0.1}{127.0.0.1:41073}{dilmrt}{ml.machine_memory=101097877504, upgraded=true, xpack.installed=true, transform.node=true, testattr=test, ml.max_open_jobs=20}, {v6.8.10-0}{GB6AeCEERkOKsU3uEWcjaw}{vN-5IJqDT6uhzW4oDb4gBw}{127.0.0.1}{127.0.0.1:40145}{dilmrt}{testattr=test, ml.machine_memory=101097877504, upgraded=true, ml.max_open_jobs=20, xpack.installed=true, transform.node=true}, {v6.8.10-1}{RDhVCXM8QsenwjJI1i0fjA}{OS1dKo8NTIyPPuJaOpLXKw}{127.0.0.1}{127.0.0.1:44217}{dilmrt}{testattr=test, ml.machine_memory=101097877504, upgraded=true, ml.max_open_jobs=20, xpack.installed=true, transform.node=true}]; discovery will continue using [[::1]:43287, 127.0.0.1:40145, [::1]:40321, 127.0.0.1:44217] from hosts providers and [{v6.8.10-2}{VOcLb7ZATs2T0VZfffUBEw}{FIQSWqWYRHG2-5gdHMidwA}{127.0.0.1}{127.0.0.1:41073}{dilmrt}{ml.machine_memory=101097877504, upgraded=true, xpack.installed=true, transform.node=true, testattr=test, ml.max_open_jobs=20}] from last-known cluster state; node term 0, last-accepted version 455 in term 0
```

```
[2020-06-02T19:24:45,421][DEBUG][o.e.a.s.m.TransportMasterNodeAction] [v6.8.10-0] timed out while retrying [cluster:monitor/health] after failure (timeout [30s])
[2020-06-02T19:24:45,421][WARN ][r.suppressed             ] [v6.8.10-0] path: /_cluster/health, params: {wait_for_status=yellow, wait_for_nodes=>=3}
```

It looks like the reason for this is that `internal:cluster/request_pre_vote` is unauthorized for `test_user`:

```
» WARN ][o.e.x.s.a.AuthorizationService] [v6.8.0-2] denying access as action [internal:cluster/request_pre_vote] is not an index or cluster action
»   ↑ repeated 39 times ↑
» WARN ][o.e.x.s.a.AuthorizationService] [v6.8.0-2] denying access as action [internal:discovery/request_peers] is not an index or cluster action
»   ↑ repeated 232 times ↑
```

```
[2020-06-02T19:24:44,726][DEBUG][o.e.c.c.PreVoteCollector ] [v6.8.10-1] TransportResponseHandler{PreVoteCollector{state=Tuple [v1=null, v2=PreVoteResponse{currentTerm=0, lastAcceptedTerm=0, lastAcceptedVersion=455}]}, node={v6.8.10-2}{VOcLb7ZATs2T0VZfffUBEw}{FIQSWqWYRHG2-5gdHMidwA}{127.0.0.1}{127.0.0.1:41073}{dilmrt}{testattr=test, ml.machine_memory=101097877504, upgraded=true, ml.max_open_jobs=20, xpack.installed=true, transform.node=true}} failed
org.elasticsearch.transport.RemoteTransportException: [v6.8.10-2][127.0.0.1:41073][internal:cluster/request_pre_vote]
Caused by: org.elasticsearch.ElasticsearchSecurityException: action [internal:cluster/request_pre_vote] is unauthorized for user [test_user]
	at org.elasticsearch.xpack.core.security.support.Exceptions.authorizationError(Exceptions.java:34) ~[?:?]
	at org.elasticsearch.xpack.security.authz.AuthorizationService.denialException(AuthorizationService.java:593) ~[?:?]
	at org.elasticsearch.xpack.security.authz.AuthorizationService.authorizeAction(AuthorizationService.java:274) ~[?:?]
	at org.elasticsearch.xpack.security.authz.AuthorizationService.maybeAuthorizeRunAs(AuthorizationService.java:231) ~[?:?]
	at org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$authorize$1(AuthorizationService.java:196) ~[?:?]
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:63) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:43) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
	at org.elasticsearch.xpack.security.authz.RBACEngine.lambda$resolveAuthorizationInfo$1(RBACEngine.java:121) ~[?:?]
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:63) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
	at org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.getRoles(CompositeRolesStore.java:238) ~[?:?]
	at org.elasticsearch.xpack.security.authz.RBACEngine.getRoles(RBACEngine.java:127) ~[?:?]
	at org.elasticsearch.xpack.security.authz.RBACEngine.resolveAuthorizationInfo(RBACEngine.java:115) ~[?:?]
	at org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:198) ~[?:?]
	at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.lambda$inbound$1(ServerTransportFilter.java:129) ~[?:?]
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:63) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$authenticateAsync$2(AuthenticationService.java:326) ~[?:?]
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$lookForExistingAuthentication$6(AuthenticationService.java:388) ~[?:?]
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lookForExistingAuthentication(AuthenticationService.java:399) ~[?:?]
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.authenticateAsync(AuthenticationService.java:323) ~[?:?]
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.access$000(AuthenticationService.java:264) ~[?:?]
	at org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:176) ~[?:?]
	at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.inbound(ServerTransportFilter.java:120) ~[?:?]
	at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler.messageReceived(SecurityServerTransportInterceptor.java:313) ~[?:?]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:72) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
	at org.elasticsearch.transport.InboundHandler$RequestHandler.doRun(InboundHandler.java:263) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:695) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_241]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_241]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_241]
```

This is unexpected: `Caused by: org.elasticsearch.ElasticsearchSecurityException: action [internal:cluster/request_pre_vote] is unauthorized for user [test_user]
`, shouldn't this be executed as system user? cc: @elastic/es-security

> This is unexpected: Caused by: org.elasticsearch.ElasticsearchSecurityException: action [internal:cluster/request_pre_vote] is unauthorized for user [test_user] , shouldn't this be executed as system user?

It should, but I can't explain it. I followed the callers of `Coordinator#startElectionScheduler` and I didn't spot any circumstances that it could run under a user context.
It started failing yesterday https://gradle-enterprise.elastic.co/s/n65kap2hjrxwg/console-log?task=:x-pack:qa:rolling-upgrade:v6.8.10%23upgradedClusterTest .

@elastic/es-distributed are you aware of anything that changed in the past couple of days that could interfere with master election scheduling?

--- a/server/src/main/java/org/elasticsearch/transport/OutboundMessage.java
+++ b/server/src/main/java/org/elasticsearch/transport/OutboundMessage.java
@@ -40,32 +40,35 @@ abstract class OutboundMessage extends NetworkMessage {
     }
 
     BytesReference serialize(BytesStreamOutput bytesStream) throws IOException {
-        storedContext.restore();
-        bytesStream.setVersion(version);
-        bytesStream.skip(TcpHeader.headerSize(version));
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            storedContext.restore();
+            bytesStream.setVersion(version);
+            bytesStream.skip(TcpHeader.headerSize(version));
 
-        // The compressible bytes stream will not close the underlying bytes stream
-        BytesReference reference;
-        int variableHeaderLength = -1;
-        final long preHeaderPosition = bytesStream.position();
+            // The compressible bytes stream will not close the underlying bytes stream
+            BytesReference reference;
+            int variableHeaderLength = -1;
+            final long preHeaderPosition = bytesStream.position();
 
-        if (version.onOrAfter(TcpHeader.VERSION_WITH_HEADER_SIZE)) {
-            writeVariableHeader(bytesStream);
-            variableHeaderLength = Math.toIntExact(bytesStream.position() - preHeaderPosition);
-        }
-
-        try (CompressibleBytesOutputStream stream = new CompressibleBytesOutputStream(bytesStream, TransportStatus.isCompress(status))) {
-            stream.setVersion(version);
-            if (variableHeaderLength == -1) {
-                writeVariableHeader(stream);
+            if (version.onOrAfter(TcpHeader.VERSION_WITH_HEADER_SIZE)) {
+                writeVariableHeader(bytesStream);
+                variableHeaderLength = Math.toIntExact(bytesStream.position() - preHeaderPosition);
             }
-            reference = writeMessage(stream);
-        }
 
-        bytesStream.seek(0);
-        final int contentSize = reference.length() - TcpHeader.headerSize(version);
-        TcpHeader.writeHeader(bytesStream, requestId, status, version, contentSize, variableHeaderLength);
-        return reference;
+            try (CompressibleBytesOutputStream stream =
+                     new CompressibleBytesOutputStream(bytesStream, TransportStatus.isCompress(status))) {
+                stream.setVersion(version);
+                if (variableHeaderLength == -1) {
+                    writeVariableHeader(stream);
+                }
+                reference = writeMessage(stream);
+            }
+
+            bytesStream.seek(0);
+            final int contentSize = reference.length() - TcpHeader.headerSize(version);
+            TcpHeader.writeHeader(bytesStream, requestId, status, version, contentSize, variableHeaderLength);
+            return reference;
+        }
     }
 
     protected void writeVariableHeader(StreamOutput stream) throws IOException {

